### ISSUE TITLE
Can we use the DHCP assigned DNS servers or an option to do so?

### ISSUE THREAD

**Issue by oitdmser** [mynodebtc/mynode#683]
**Is your content request related to a problem? Please describe.**
Currently MyNode disables openresolvconf using resolvconf=NO in /etc/resolvconf.conf and statically assigns DNS servers 1.1.1.1, 208.67.222.222, 8.8.8.8, and 8.8.4.4 using /usr/bin/mynode_startup.sh. 

I am not comfortable with using Google and Cloudfare DNS servers. I use a local Pi-Hole for DNS, which in turn uses my preferred non-Google, non-Cloudflare  DNS servers. 

**Describe the solution you'd like**
It would be cool if MyNode could use the DNS server assigned by DHCP (in my case my local Pi-Hole).

**Additional context**
This could be accomplished in a couple ways.

1. enable openresolvconf and disable static DNS assignments in mynodestartup.sh2. 
2. add a UI option to manually assign DNS
3. add a UI option to manually assign DNS with a check-box to "use DHCP DNS"

- [X] Mark this if you searched the existing list of open/closed issue and did not find anything similar

**Comment by tehelsper** [comment]
It already should be using your local DNS servers and preferring them over the static ones by appending static to the list. See /etc/dhcp/dhclient.conf. They were added as a fallback since some early users had issues resolving hostnames. DNS is managed via network manager.

Try running this command:

nmcli dev show | grep DNS

**Comment by oitdmser** [comment]
`nmcli dev show | grep DNS` returns only one entry, my local DNS server, so that part appears correct. However, a cat /etc/resolv.conf shows the following:

```
# Added at myNode startup
nameserver 1.1.1.1
nameserver 208.67.222.222
nameserver 8.8.8.8
nameserver 8.8.4.4
```

my resolvconf.conf is as follows:

```
# Configuration for resolvconf(8)
# See resolvconf.conf(5) for details

resolvconf=NO
```

I deleted resolv.conf and reboot, and the same thing happens.

**Comment by tehelsper** [comment]
Yeah, they will always appear, but I don't think they are in use.

Try "dig google.com" or "nslookup github.com" and see which servers are used.

**Comment by oitdmser** [comment]
Looks like it's using the first entry in resolv.conf:

```
; <<>> DiG 9.11.5-P4-5.1+deb10u7-Debian <<>> google.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 45838
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;google.com.			IN	A

;; ANSWER SECTION:
google.com.		83	IN	A	142.251.33.78

;; Query time: 24 msec
;; SERVER: 1.1.1.1#53(1.1.1.1)
;; WHEN: Fri Jul 22 17:40:44 CDT 2022
;; MSG SIZE  rcvd: 55


nslookup google.com
Server:		1.1.1.1
Address:	1.1.1.1#53

Non-authoritative answer:
Name:	google.com
Address: 142.251.33.78
Name:	google.com
Address: 2607:f8b0:400a:806::200e



```

**Comment by oitdmser** [comment]
I discovered that NetworkManager is in fact creating /run/NetworkManager/resolv.conf, but because /etc/resolv.conf is not symlinked that doesn't get propagated. I created a symlink to /run/NetworkManger/resolv.conf and it works now.

**Comment by oitdmser** [comment]
Something is definitely up with the mynode config. After a reboot resolv.conf looks like this:

```
# Added at myNode startup
nameserver 1.1.1.1
nameserver 208.67.222.222
nameserver 8.8.8.8
nameserver 8.8.4.4
```
However /etc/resolv.conf is still a symlink:
```
$ ls -la resol*
lrwxrwxrwx 1 root root   31 Jul 22 18:06 resolv.conf -> /run/NetworkManager/resolv.conf
```

I reset Network Manager multiple times with systemctl restart network-manager and both /run/NetworkManager/resolv.conf and /etc/resolv.conf are the same as above.

**Comment by oitdmser** [comment]
FWIW I added `rc-manager=symlink` to the [main] section of /etc/NetworkManager/NetworkManager.conf and reloaded network manager. Both resolv.conf files now populate. After a reboot  I see

```
# Generated by NetworkManager
search mydomain.org
nameserver x.y.z.a

# Added at myNode startup
nameserver 1.1.1.1
nameserver 208.67.222.222
nameserver 8.8.8.8
nameserver 8.8.4.4
```

this seems to be the intended behavior.

**Comment by tehelsper** [comment]
Thanks! For some reason I thought that behavior was already occurring, but it must not have been without the extra line in NetworkManager.conf. I will add this in the next release.

**Comment by oitdmser** [comment]
one issue I have with this setup is that DNS requests still leak out to the foreign name servers. My router logs DNS queries, and when I have the above config with my nameserver at the top of the list, the router still sees some queries going out to 208.67.222.222 from mynode.

Is it possible to eliminate the whole "#Added at myNode startup" section if there is already a name server in resolve.conf?

**Comment by oitdmser** [comment]
Update. A tcpdump reveals some process is querying the opendns server 208.67.222.222 for myip.opendns.com. Is this the method myNode uses to determine it's IP? It seems that 208.67.222.222 is hardcoded, which is bad practice. Can that be remedied?

**Comment by tehelsper** [comment]
Yeah, I just published a new beta that adds a toggle on the settings page to remove the backup DNS servers. If you want to upgrade, you can test it out.

It worked for me, but I did have to force a DHCP release/renew before the DNS settings were updated. I imagine it would also work if you waited for a new renew cycle.

**Comment by oitdmser** [comment]
that's awesome!

My node runs as a production node so I don't want to risk a beta, but I'll keep an eye out when I upgrade. Will that toggle show up in v0.3.0 or v0.3.1?

**Comment by tehelsper** [comment]
It will be available in v0.3.1.

**Comment by tensor5g** [comment]
This setting doesn't appear to be working for me. I am monitoring traffic from my firewall and my mynode server is reaching out to  87.236.195.203 on port 53 constantly. It's the only device on my network that isn't using my internal DNS server.

**Comment by tehelsper** [comment]
That IP isn't set anywhere within our software, but it could be one of the apps directly reaching out. Can you tell which app is attempting to use that DNS server?

**Comment by tensor5g** [comment]
You are probably right about one of the apps doing it, although I tried disabling them one by one and the traffic did not stop. Also disabled all of them, same result. Maybe it needs a restart after disabling the app?
